tests/test_query.py

```diff
@@ -83,12 +70,12 @@ def mock_httpx_post(**kwargs):
 
 
 def test_get_all(test_app, mock_successful_get, monkeypatch):
-    """Given no input for the sex parameter, returns a 200 status code and a non-empty list of results (should correspond to all subjects in graph)."""
+    """Given no input for the sex parameter, returns a 200 status code and at least one dataset with subjects (should correspond to all subjects in graph)."""
 
     monkeypatch.setattr(crud, "get", mock_successful_get)
     response = test_app.get("/query/")
     assert response.status_code == 200
-    assert response.json() != []
+    assert 0 not in [i["num_matching_subjects"] for i in response.json()]
```

Comment:
I like the more explicit test. But I find this assertion confusing. The `"num_matching_subjects"` thing is coming from `get`, but we know that this is mocked out. So you are essentially just asserting that the mocked data fixture from above has a key for `num_matching...` and that its value is not 0.

In this case I think it makes more sense to keep the previous assertion because that just checks that something comes back out of the app. We can also talk through the tests a bit more in https://github.com/neurobagel/bagelapi/issues/48 or something related.

Reply:
I don't think having the old version makes sense in the current implementation, the response is going to be a list either way. We wanted to differentiate between a (happy) test with no matching data and one with matching data e.g.

https://github.com/neurobagel/bagelapi/blob/aab9571e865f1822b50451e915203c2033bbb009/tests/test_query.py#L353 

Here we expect no matching subjects (the old empty list) as a response so we check if the number of matching subjects is zero for the only dictionary in the returned list.
Another reason I had in mind was to in a way document our intention with the current implementation and our expectation of how the API would respond i.e. a list containing dictionaries (corresponding to each dataset) none of which have 0 `num_matching_subjects`.

Reply:
OK, I understand. But even with the old test you aren't checking whether the result is a list, but whether it's an empty list. So before the test read like: 

> "get status 200 and a non-empty result list".

Now the test reads 

> "get status 200 and also make sure that none of the objects in the list have `0` as their `num_matching_subjects` attribute". 

But I think there isn't really a way for that second statement to ever be wrong (if the API runs at all). You mock out the `get` and make it return a list with a known content. And the API essentially takes this list and without doing anything else returns it to your test client. So in other words: what would I learn from this test failing?

Reply:
> OK, I understand. But even with the old test you aren't checking whether the result is a list, but whether it's an empty list

Yea let me correct myself currently, the result is going to be a non-empty list either way so the old assertion doesn't differentiate between an empty response and a non-empty one.

> But I think there isn't really a way for that second statement to ever be wrong (if the API runs at all). You mock out the get and make it return a list with a known content. And the API essentially takes this list and without doing anything else returns it to your test client. So in other words: what would I learn from this test failing?

I see your point but I feel like that's the case with (tests from the) previous implementation. We always mock the `get` function and assert over the return from the mock and as you mentioned, the API essentially takes that and returns it to the test client without doing anything so there isn't really a way for the second statement to ever be wrong (as long as the API runs with no issue).

I look at the second assertion statement as a way for us to document the expected behavior of the API and our decisions regarding implementation e.g. when we have a valid but not available image modality in the query of the API we decided we would let it pass and return an empty list which in the current implementation would be a list with a dictionary pointing to 0 matched subjects and its test will reflect that.

If I'm not mistaken @alyssadai and I discussed whether it makes sense to have a second assert statement back when we were first setting up tests. I think we can circle back and discuss it in #48.

Reply:
OK, makes sense. I thought before we doing some parsing on top of the mocked get return so at least that was being tested. We shouldn't hold up this PR over that question, let's discuss the tests somewhere else and then make a decision.